execution/gethexec/express_lane_service.go

```diff
@@ -0,0 +1,456 @@
+					// by sequence number must be reset otherwise old messages would be replayed.
+					es.messagesBySequenceNumber = make(map[uint64]*timeboost.ExpressLaneSubmission)
+					es.Unlock()
+				}
+				fromBlock = toBlock
+			}
+		}
+	})
+}
+
+func (es *expressLaneService) currentRoundHasController() bool {
+	currRound := timeboost.CurrentRound(es.initialTimestamp, es.roundDuration)
+	control, ok := es.roundControl.Get(currRound)
+	if !ok {
+		return false
+	}
+	return control.controller != (common.Address{})
+}
+
+func (es *expressLaneService) isWithinAuctionCloseWindow(arrivalTime time.Time) bool {
+	// Calculate the next round start time
+	elapsedTime := arrivalTime.Sub(es.initialTimestamp)
+	elapsedRounds := elapsedTime / es.roundDuration
+	nextRoundStart := es.initialTimestamp.Add((elapsedRounds + 1) * es.roundDuration)
+	// Calculate the time to the next round
+	timeToNextRound := nextRoundStart.Sub(arrivalTime)
+	// Check if the arrival timestamp is within AUCTION_CLOSING_DURATION of TIME_TO_NEXT_ROUND
+	return timeToNextRound <= es.auctionClosing
+}
+
+// Sequence express lane submission skips validation of the express lane message itself,
+// as the core validator logic is handled in `validateExpressLaneTx“
+func (es *expressLaneService) sequenceExpressLaneSubmission(
+	ctx context.Context,
+	msg *timeboost.ExpressLaneSubmission,
+) error {
+	es.Lock()
+	defer es.Unlock()
+	// Although access to roundControl by itself is thread-safe, when the round control is transferred
+	// we need to reset roundControl and messagesBySequenceNumber atomically, so the following access
+	// must be within the lock.
+	control, ok := es.roundControl.Get(msg.Round)
+	if !ok {
+		return timeboost.ErrNoOnchainController
+	}
+
+	// Check if the submission nonce is too low.
+	if msg.SequenceNumber < control.sequence {
+		return timeboost.ErrSequenceNumberTooLow
+	}
+	// Check if a duplicate submission exists already, and reject if so.
+	if _, exists := es.messagesBySequenceNumber[msg.SequenceNumber]; exists {
+		return timeboost.ErrDuplicateSequenceNumber
+	}
+	// Log an informational warning if the message's sequence number is in the future.
+	if msg.SequenceNumber > control.sequence {
+		log.Info("Received express lane submission with future sequence number", "SequenceNumber", msg.SequenceNumber)
+	}
+	// Put into the sequence number map.
+	es.messagesBySequenceNumber[msg.SequenceNumber] = msg
+
+	for {
+		// Get the next message in the sequence.
+		nextMsg, exists := es.messagesBySequenceNumber[control.sequence]
+		if !exists {
+			break
+		}
+		if err := es.transactionPublisher.PublishTimeboostedTransaction(
+			ctx,
+			nextMsg.Transaction,
+			msg.Options,
+		); err != nil {
+			// If the tx failed, clear it from the sequence map.
+			delete(es.messagesBySequenceNumber, msg.SequenceNumber)
+			return err
+		}
+		// Increase the global round sequence number.
+		control.sequence += 1
+	}
+	es.roundControl.Add(msg.Round, control)
+	return nil
+}
```

Comment:
There's a bug here as an invalid transaction could essentially block all the txs with future sequence numbers as this tx would itself error out and returns an error (to possibly an incorrect sender) but also deletes itself from the map thus blocking all future txs until a valid tx with this seq num is posted!

```suggestion

        var relevantErr error
	for {
		// Get the next message in the sequence.
		nextMsg, exists := es.messagesBySequenceNumber[control.sequence]
		if !exists {
			break
		}
		if err := es.transactionPublisher.PublishTimeboostedTransaction(
			ctx,
			nextMsg.Transaction,
			msg.Options,
		); err != nil && msg.SequenceNumber == nextMsg.SequenceNumber {
			// If the tx failed and if its the currently submitted one we track this error else continue
			relevantErr = err
		}
		delete(es.messagesBySequenceNumber, nextMsg.SequenceNumber)
		// Increase the global round sequence number.
		control.sequence += 1
	}
	es.roundControl.Add(msg.Round, control)
	return relevantErr
}
```

Above suggestion also takes into account Tsahi's review of not keeping already processed txs

Reply:
if a tx which was not the currently submitted one fails, we should give extra information to the sender so that they can retry with the correct sequence number